Refactor board logic, add play, captures, suicide detection, SGF printer

espadrine opened the pull request:
I somewhat tested that play works correctly by adding a non-trivial test, and checking the result of the first face-to-face AlphaGo vs. Ke Jie match:

![image](https://user-images.githubusercontent.com/100689/26859097-b3febbc8-4b35-11e7-9dec-b7d39758a715.png)

@jankeromnes, could you have a look? In particular, f44836a, which is inspired by #1.

jankeromnes committed:
Implement board.play(x, y) with captures and suicides

Co-written with Thaddée Tyl.

espadrine committed:
Add Intersection objects on the Go board

Each intersection has detailed information about that point on the board.

espadrine committed:
Test multi-group capture

espadrine committed:
Add a board to string displayer

espadrine committed:
Add primitives to run an SGF file

espadrine commented on src/board.js:
```diff
@@ -9,22 +9,232 @@
+
+      if (capturedEnemyGroups.length === 0) {
+        // We are not capturing enemy stones. Are we committing suicide?
+        var isDirectlySurroundedByEnemy = enemySurroundingGroups.length === 4;
```
This doesn't look right, I'll fix it. There can be a single enemy group surrounding the point where we try to place a stone.

espadrine replied:
Actually, it's correct. There can be duplicate groups in `enemySurroundingGroups`, as there are always four groups in `surroundingGroups`.
Will add a test.

espadrine replied:
Nice catch! I'll test that as well, and fix it.

espadrine replied:
Funnily enough, it actually works! While the `isDirectlySurroundedByEnemy` part is wrong, the `isKillingOwnGroup` check actually yields the correct result even when there are no own surrounding groups.

jankeromnes commented:
Oh wow, this came in fast! I didn't expect such a feature-complete pull request so soon. 💯

jankeromnes reviewed (changes requested):
I did a first quick pass through the code, it's amazing to see it take shape! 😄

On a general note, a lot of the game logic seems convoluted because of many `filter` calls that pass through arrays multiple times and create many temporary arrays. I believe it could be simplified a lot by doing all the computation in one for loop through all the neighbors (and a few post-loop instructions).

Please request a follow-up review so that I have more time to review the game logic.

Thanks!

espadrine commented on src/board.js:
```diff
@@ -9,22 +9,232 @@
+    this.ownLibertiesChange = 0;
+    this.enemyLibertiesChange = 0;
+    this.capturesFromMove = 0;  // Number of enemy stones it would capture.
+    this.selfAtariFromMove = 0;  // Number of own stones it would capture.
```
The Nature AlphaGo paper is unclear, but my guess is that this is the algorithm they copied: [Pachi self-atari](https://github.com/pasky/pachi/blob/master/tactics/selfatari.h#L59).

espadrine commented on src/board.js:
```diff
@@ -9,22 +9,232 @@
+    this.enemyLibertiesChange = 0;
+    this.capturesFromMove = 0;  // Number of enemy stones it would capture.
+    this.selfAtariFromMove = 0;  // Number of own stones it would capture.
+    this.sensibleMove = true;  // ie, legal and does not fill its own eyes.
```
Here, too, the AlphaGo paper is unclear. However, they clearly read the Pachi source a lot, so my guess is that they use its [big eye algorithm](https://github.com/pasky/pachi/blob/1311539485307145ae4886115eeddfeea41f85a7/tactics/dragon.c#L293).

espadrine commented:
Thanks for the review!

I purposefully did not make any effort to be fast, so that we could start with something that was correct, and try to improve from there. I'd like performance optimizations to be in follow-up patches, if you don't mind. And there's definitely a ton of low-hanging fruits ☺

espadrine commented on src/board.js:
```diff
@@ -9,22 +9,232 @@
+
+  function rowNameFromNum(number) {
+    if (number < 26) { return String.fromCharCode(97 + number); }
+    else { return String.fromCharCode(65 + number - 26); }
```
Going from `z` to `A` is defined in SGF. Nothing is defined beyond `Z`. I can add a check to refuse positions beyond 51.

espadrine commented on src/board.js:
```diff
@@ -9,22 +9,232 @@
+    surrounding: function(x, y) {
+      var top    = (y > 0)?             this.get(x    , y - 1): undefined;
+      var right  = (x < this.size - 1)? this.get(x + 1, y    ): undefined;
+      var bottom = (y < this.size - 1)? this.get(x    , y + 1): undefined;
```
I think not: the maximum `y` is `this.size - 2`.

espadrine commented on src/board.js:
```diff
@@ -9,22 +9,232 @@
+      var self = this;
+      var point = self.get(x, y);
+      var color = self.nextPlayingColor;
+      if (point === undefined || point.color !== Board.EMPTY) { return false; }
```
Good point. I feel like we should place these checks in `board.get(x, y)`.

espadrine commented on src/board.js:
```diff
@@ -9,22 +9,232 @@
+    this.color = Board.EMPTY;
+    this.points = new Set();
+    this.liberties = new Set();  // Set of points.
+    points.forEach(this.addPoint.bind(this));
```
It's a perf improvement; let's put it as a follow-up.

espadrine commented on src/board.js:
```diff
@@ -9,22 +9,232 @@
+          points.push(point);
+        });
+      });
+      return new Group(this, points);
```
This is a perf improvement; let's put it in a follow-up.

espadrine commented on src/board.js:
```diff
@@ -9,22 +9,232 @@
+    },
+
+    surrounding: function(x, y) {
+      var top    = (y > 0)?             this.get(x    , y - 1): undefined;
```
I feel it could be useful, even though it is not currently in use. It does carry the meaning that we are on a given edge, which we wouldn't know otherwise.

But given that we currently don't use that fact, I followed your advice. We can change it back if we ultimately need it.

espadrine commented on src/board.js:
```diff
@@ -9,22 +9,232 @@
+      });
+      var capturedEnemyGroups = enemySurroundingGroups.filter(function(group) {
+        return group.liberties.size === 1;
+      });
```
We can do performance improvements in a follow-up.

espadrine commented:
Apart from performance improvements, I fixed the commits in accordance to your reviews. Do you want to have a last look before I merge?

If you want, you can do the performance improvement patches yourself; I added you as a collaborator.
